### PR TITLE
Avoid unneeded refresh with concurrent realtime gets

### PR DESCRIPTION
**Pull request by dnhatn**
This change should reduce refreshes for a use-case where we perform multiple realtime gets at the same time on an active index. Currently, we only call [refresh](https://github.com/elastic/elasticsearch/blob/ae5b876a0726cb88818525c5e79e4a57abb8251f/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L674) if the index operation is still on the versionMap. However, at the time we call refresh, that operation might be already or will be included in the latest reader. Hence, we do not need to refresh. Adding another lock here is not an issue as the refresh is already sequential.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -671,7 +671,7 @@ public GetResult get(Get get, BiFunction<String, SearcherScope, Engine.Searcher>
                             trackTranslogLocation.set(true);
                         }
                     }
-                    refresh("realtime_get", SearcherScope.INTERNAL, true);
+                    refreshIfNeeded("realtime_get", versionValue.seqNo);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can you add an assertion that `versionValue.seqNo >= 0`?

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 2]
```diff
@@ -6102,4 +6103,57 @@ private void runTestDeleteFailure(
...
     }
 
+    public void testRealtimeGetOnlyRefreshIfNeeded() throws Exception {
+        final AtomicInteger refreshCount = new AtomicInteger();
+        final ReferenceManager.RefreshListener refreshListener = new ReferenceManager.RefreshListener() {
+            @Override
+            public void beforeRefresh() {
+
+            }
+            @Override
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: add newline

**Review by ywelsch — APPROVED** [review]
LGTM

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by dnhatn** [comment]
The test failure is tracked at https://github.com/elastic/elasticsearch/issues/43673

**Comment by dnhatn** [comment]
@ywelsch and @henningandersen thanks for reviewing.

**Pull request merged**